### PR TITLE
feat: linked providers

### PR DESCRIPTION
**Pull request by shahargl**
close https://github.com/keephq/keep/issues/935

### CODE REVIEW

File: keep/api/core/db.py [hunk 1]
```diff
@@ -1388,3 +1388,17 @@ def update_key_last_used(
         tenant_api_key_entry.last_used = datetime.utcnow()
         session.add(tenant_api_key_entry)
         session.commit()
+
+
+def get_linked_providers(tenant_id: str) -> List[Provider]:
```
Comment (talboren) [inline keep/api/core/db.py:1393]:
```suggestion
def get_linked_providers(tenant_id: str):
```

File: keep-ui/app/providers/provider-tile.tsx [hunk 2]
```diff
@@ -101,14 +108,28 @@ export default function ProviderTile({ provider, onClick }: Props) {
...
+                <Subtitle className="group-hover:hidden">
+                  Last alert: {moment(provider.last_alert_received).fromNow()}
+                </Subtitle>
+              )
+            }
+            {
+              provider.linked && (
+                <Subtitle className="group-hover:hidden">
+                  Name: {provider.}
+                </Subtitle>
```
Comment (talboren) [inline keep-ui/app/providers/provider-tile.tsx:127]: This seems to fail the build

File: keep-ui/app/providers/providers-tiles.tsx [hunk 3]
```diff
@@ -112,8 +114,9 @@ const ProvidersTiles = ({
   return (
     <div>
       <Title className="mb-2.5">
-        {installedProvidersMode ? "Installed Providers" : "Available Providers"}
+        {installedProvidersMode ? "Installed Providers" : linkedProvidersMode ? "Linked Providers" : "Available Providers"}
```
Comment (talboren) [inline keep-ui/app/providers/providers-tiles.tsx:117]: Perhaps also add a small question mark icon that explains what are Linked Providers?

File: keep/api/core/db.py [hunk 4]
```diff
@@ -1388,3 +1388,17 @@ def update_key_last_used(
...
+    with Session(engine) as session:
+        # extract all providers from the alerts that are linked to the tenant
+        providers = session.exec(
+            select(Alert.provider_type, func.max(Alert.timestamp))
```
Comment (talboren) [inline keep/api/core/db.py:1397]: Also get the provider name if it's available?

**Review by talboren — CHANGES_REQUESTED** [review]
A few small comments but LGTM!

File: keep-ui/app/providers/page.client.tsx [hunk 5]
```diff
@@ -17,11 +17,11 @@ import { LayoutContext } from "./context";
 import { toast } from "react-toastify";
 import { updateIntercom } from "@/components/ui/Intercom";
 import { useRouter } from "next/navigation";
-import { Callout } from "@tremor/react";
 
 export const useFetchProviders = () => {
   const [providers, setProviders] = useState<Provider[]>([]);
   const [installedProviders, setInstalledProviders] = useState<Provider[]>([]);
+  const [linkedProviders, setLinkedProviders] = useState<Provider[]>([]); // Added state for linkedProviders
```
Comment (asharonbaltazar) [inline keep-ui/app/providers/page.client.tsx:24]: Should comments be a little more descriptive? 👀

**Review by asharonbaltazar — COMMENTED** [review]
Do you have Prettier installed on your IDE? Some of the code doesn't seem to be formatted

File: keep-ui/app/providers/page.client.tsx (hunk 5 above)
Comment (talboren) [reply keep-ui/app/providers/page.client.tsx:24]: 🤖

File: keep-ui/app/providers/page.client.tsx [hunk 6]
```diff
@@ -72,54 +72,39 @@ export const useFetchProviders = () => {
...
-    setProviders(fetchedProviders);
-  }
+          validatedScopes: provider.validatedScopes ?? {},
+        })
+      );
+
+      const fetchedProviders = data.providers.map((provider) => ({
+        ...defaultProvider,
+        ...provider,
+        installed: provider.installed ?? false,
```
Comment (talboren) [inline keep-ui/app/providers/page.client.tsx:88]: not that it's that important but aren't "linked providers" always "installed"?

File: keep/api/core/db.py [hunk 7]
```diff
@@ -1438,3 +1439,89 @@ def update_key_last_used(
...
+    return providers
+
+
+def get_provider_distribution(tenant_id: str) -> dict:
+    """Returns hits per hour and the last alert timestamp for each provider, limited to the last 24 hours."""
+    with Session(engine) as session:
+        twenty_four_hours_ago = datetime.utcnow() - timedelta(hours=24)
+        time_format = "%Y-%m-%d %H"
+
+        if session.bind.dialect.name in ["mysql", "postgresql"]:
```
Comment (talboren) [inline keep/api/core/db.py:1472]: maybe just leave a todo `# mssql`

**Review by talboren — APPROVED** [review]
LGTM

**Pull request merged**